gt/main.c

```diff
@@ -16,11 +16,254 @@
+			 * capabilities, and when each decision expires.
+			 */
+
+			/* TODO Reply the policy decision to GK-GT unit. */
+		}
+
+		/* Send burst of TX packets, to second port of pair. */
+		num_tx_succ = rte_eth_tx_burst(port, tx_queue,
+			rx_bufs, num_rx);
+
+		/*
+		 * XXX Do something better here!
+		 * For now, free any unsent packets.
+		 */
+		if (unlikely(num_tx_succ < num_rx)) {
+			for (i = num_tx_succ; i < num_rx; i++)
+				rte_pktmbuf_free(rx_bufs[i]);
+		}
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gt: the GT block at lcore = %u is exiting\n", lcore);
+
+	return gt_conf_put(gt_conf);
+}
+
+struct gt_config *
+alloc_gt_conf(void)
+{
+	return rte_calloc("gt_config", 1, sizeof(struct gt_config), 0);
+}
+
+static int
+cleanup_gt(struct gt_config *gt_conf)
+{
+	rte_free(gt_conf->instances);
+	rte_free(gt_conf->lcores);
+	rte_free(gt_conf);
+
+	return 0;
+}
 
 int
-run_gt(__attribute__((unused)) const struct gt_config *gt_conf)
+gt_conf_put(struct gt_config *gt_conf)
 {
-	/* TODO Initialize and run GT functional block. */
+	/*
+	 * Atomically decrements the atomic counter (v) by one and returns true 
+	 * if the result is 0, or false in all other cases.
+	 */
+	if (rte_atomic32_dec_and_test(&gt_conf->ref_cnt))
+		return cleanup_gt(gt_conf);
+
+	return 0;
+}
+
+static int
+init_gt_instances(struct gt_config *gt_conf)
+{
+	int i;
+	int ret;
+
+	/* Set up queue identifiers now for RSS, before instances start. */
+	for (i = 0; i < gt_conf->num_lcores; i++) {
+		unsigned int lcore = gt_conf->lcores[i];
+		struct gt_instance *inst_ptr = &gt_conf->instances[i];
+
+		ret = get_queue_id(&gt_conf->net->front, QUEUE_TYPE_RX, lcore);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER, "gt: cannot assign an RX queue for the front interface for lcore %u\n",
+				lcore);
+			goto out;
+		}
+		inst_ptr->rx_queue = (uint16_t)ret;
+
+		ret = get_queue_id(&gt_conf->net->front, QUEUE_TYPE_TX, lcore);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER, "gt: cannot assign a TX queue for the front interface for lcore %u\n",
+				lcore);
+			goto out;
+		}
+		inst_ptr->tx_queue = (uint16_t)ret;
+	}
+
+	ret = 0;
+
+out:
+	return ret;
+}
+
+static int
+gt_stage1(void *arg)
+{
+	int ret;
+	struct gt_config *gt_conf = arg;
+
+	gt_conf->instances = rte_calloc(__func__, gt_conf->num_lcores,
+		sizeof(struct gt_instance), 0);
+	if (gt_conf->instances == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	ret = init_gt_instances(gt_conf);
+	if (ret < 0)
+		goto  instance;
+
+	goto out;
+
+instance:
+	rte_free(gt_conf->instances);
+	gt_conf->instances = NULL;
+	rte_free(gt_conf->lcores);
+	gt_conf->lcores = NULL;
+out:
+	return ret;
+}
+
+static int
+gt_stage2(void *arg)
+{
+	struct gt_config *gt_conf = arg;
+	return gt_setup_rss(gt_conf);
+}
+
+int
+run_gt(struct net_config *net_conf, struct gt_config *gt_conf)
+{
+	int ret, i;
+
+	if (net_conf == NULL || gt_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	gt_conf->net = net_conf;
+
+	if (gt_conf->num_lcores <= 0) {
+		ret = -1;
+		goto out;
```

Comment:
The only reason the line `gt_conf->net = net_conf;` comes early in `run_gk()` is because `gk_conf->num_lcores <= 0` is a success condition, that is, if there's no lcores to bootstrap, there's nothing to do. Unless you have a reason to not assume that, I'd follow `run_gk()`'s example to keep the code uniform.